File: pypp_cli/test_dir/cpp/src/exceptions/test_all.cpp
#include "exceptions/test_all.h"
#include "exceptions/common.h"
#include "exceptions/custom_exceptions.h"
#include "exceptions/exception.h"
#include "py_list.h"
#include "py_str.h"
#include "pypp_util/print.h"
#include "pypp_util/to_py_str.h"
#include <string>

namespace me {
void test_all_exceptions_fn() {
    pypp::print(pypp::PyStr("ALL EXCEPTIONS TEST RESULTS:"));
    pypp::PyList<pypp::Exception> exceptions(
        {pypp::Exception(pypp::PyStr("test")),
         pypp::ValueError(pypp::PyStr("test"))});
    for (const auto &exc : exceptions) {
        try {
            throw exc;
        } catch (const pypp::Exception &pypp_e) {
            std::string e = pypp_e.what();
            pypp::print(pypp::PyStr("caught exception: ") + pypp::str(e));
        }
    }
    try {
        throw me::ChildException(pypp::PyStr("test"));
    } catch (const me::ChildException &pypp_e) {
        std::string e = pypp_e.what();
        pypp::print(pypp::PyStr("caught exception: ") + pypp::str(e));
    }
}

} // namespace me

File: pypp_cli/test_dir/python/src/exceptions/test_all.py
from exceptions.custom_exceptions import ChildException


def test_all_exceptions_fn():
    print("ALL EXCEPTIONS TEST RESULTS:")
    # TODO: add the others. But first they all need to inherit from Exception
    exceptions: list[Exception] = [
        Exception("test"),
        ValueError("test"),
        IndexError("test"),
        KeyError("test"),
        AssertionError("test"),
        RuntimeError("test"),
        NotImplementedError("test"),
        OSError("test"),
        FileNotFoundError("test"),
        NotADirectoryError("test"),
        PermissionError("test"),
        FileExistsError("test"),
    ]
    for exc in exceptions:
        try:
            raise exc
        except Exception as e:
            print("caught exception: " + str(e))

    # also test importing another exception
    try:
        raise ChildException("test")
    except ChildException as e:
        print("caught exception: " + str(e))
